src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -427,11 +433,22 @@ public static final class PlaceholderTask implements ContinuedTask, Serializable
 
         private Object readResolve() {
             if (cookie != null) {
-                synchronized (runningTasks) {
-                    runningTasks.put(cookie, new RunningTask());
-                }
+                RunningTasks.get().withRunningTasks(runningTasks -> {
+                    // If Jenkins stops while this step is resuming, there may be a PlaceholderTask in the queue as
+                    // well as in program.dat for the same step. We want to make sure not to create a second task to
+                    // avoid race conditions, so we use putIfAbsent.
+                    // TODO: This helps for runningTasks, but other fields like `stopping` may still be problematic.
+                    // Should we refactor things to guarantee that the relevant state is a singleton? For example,
+                    // introduce a PlaceholderTasksAction that holds a map of a new PlaceholderTaskState class, which
```

Comment:
An action on a `FlowNode` you mean?

Reply:
I was thinking on the `Run`, but yes perhaps it could be on the `FlowNode` in which case the action would be simpler at the cost of needing a live execution to find the task, but that might be fine here.